### ISSUE TITLE
`autocapitalize` type is incorrect according to MDN

### ISSUE THREAD

**Issue by michaelwarren1106** [microsoft/TypeScript#54372]
# lib Update Request

In lib/dom.d.ts

https://github.com/microsoft/TypeScript/blob/main/src/lib/dom.generated.d.ts#L9992

the type for `HTMLElement.autocapitalize` is incorrectly typed as `string`. According to [the MDN article](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autocapitalize), the `autocapitalize` attribute/property can only be set to certain keywords.

Therefore, the internal lib dom type should be:

`autocapitalize: 'off' | 'none' | 'on' | 'sentences' | 'words' | 'characters';`

## Sample Code

This issue affects proper typing in web components that extend HTMLElement that attempt to correctly type `autocapitalize`. TS errors because `string` and a union type are not compatible. Every reference to a web component class extending HTMLElement with a proper type for `autocapitalize` errors incorrectly.

## Documentation Link

[MDN article on autocapitalize](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autocapitalize)

**Comment by Peeja** [comment]
I believe [enumerated attributes](https://developer.mozilla.org/en-US/docs/Glossary/Enumerated) are typically typed as `string` because of the level of correctness they're meant to check. In particular, it's not an _error_ to assign a random string:

> Each enumerated attribute has a default value for when the attribute is present without a value (the value is missing), and a default value for when the attribute is assigned an invalid value.

Testing in Chrome:

```js
d = $("div") // random div off the page
d.autocapitalize // ''
d.autocapitalize = "foobar"
d.autocapitalize // 'sentences'
```

So:

* It's legal (though not useful) to *assign* any string to the property.
* Assigning an invalid value causes the browser to default to a particular valid value.
* (At least in Chrome:) Reading the property will always give you the effective value, and therefore one of the valid values—and this is notably not reflected in the types.
* `''` is a valid value as well.

You mention that:

> TS errors because `string` and a union type are not compatible. Every reference to a web component class extending HTMLElement with a proper type for `autocapitalize` errors incorrectly.

Can you reproduce the problem you're seeing in a Playground? I've been unable to make it cause an error. [This much seems to work fine](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIFcDOYD2BbAUQBsI8JxkIAPSEAE02QAkAVAWQBljTyxkBvAFDIRyOOlwI4AB2Bg4RYAC8IALmQByHDBgbkAH00gcICHsNaQ5zZl7kkmaxoDuOKAycIAFnCiJIUI4A3IIAvoKCMOggCGDAJsh0OADK+BBgXqAA5gDqcl4AgiCsnNxk4AAUaszsXCTlYACUAuGCdBAIRL4oCCbYyAhYuIT1vOoY2PhlvCFJqWQZ2XkZRSV1PJWDkyMbTUA):

```ts
interface CustomElement extends HTMLElement {
    autocapitalize: 'off' | 'none' | 'on' | 'sentences' | 'words' | 'characters';
}

function doSomethingWithAnHTMLElement(e: HTMLElement) {}

declare const customElement: CustomElement;
doSomethingWithAnHTMLElement(customElement)
```

**Comment by Peeja** [comment]
> Since `autocapitalize` has a fallback it is technically optional isn't it?

No, I don't believe the spec allows `autocapitalize` to be *undefined*. It does allow it to be `''`, though, which yields the default behavior.

Oddly, when I set a `div`'s `autocapitalize` to `undefined` in Chrome, it turns it into…`'sentences'`. Weird. I'd venture to say that behavior is, well, undefined. 😄

> Also regarding the level of error. IMO Typescript should be checking this kind of thing even if the browser will reset the value internally to a valid one if incorrectly set. Its "legal" only because the browser basically wont let you set it incorrectly. But shouldn't TS warn at least before trying to set it incorrectly in the first place?

I think that's a reasonable line to draw. It's just not the line that _has_ been drawn so far. Moving that line would mean updating these for every enumerated attribute in the DOM spec, _and_ would be a breaking change for any code relying on the (admittedly sketchy) defaulting behavior. That's a pretty big thing to ask. There's an argument to make for it, it's just bigger than this issue looked like to begin with.

**Comment by fatcerberus** [comment]
Note that `exactOptionalPropertyTypes` prevents *setting* the value of an optional property to `undefined`, but wouldn't prevent `delete`'ing it, *and* you'd get an annoying `string | undefined` (which can't actually happen, it's always a string) any time you tried to read the current value.

**Comment by michaelwarren1106** [comment]
ultimately i’m trying to figure out the “correct” type for a custom element thats just gonna pass this attribute/prop value down from a custom element host to a native input in shadow dom. 

should it just be ! instead of ? to get rid of the error and provide an optional prop from the custom element level

**Comment by Peeja** [comment]
@michaelwarren1106 I think that would just be `string`. Or the union, if you want to be more restrictive than the lib definition. If you specifically want it to be optional, you can accept and handle `undefined` yourself and only pass it along when it's defined. Does that work for your code?

**Comment by michaelwarren1106** [comment]
i’m already doing that, but in a way that typescript
doesn’t know about. i just made it non-null with ! and TS seems happy. /shrug

**Comment by RyanCavanaugh** [comment]
We generally accept changes of this nature even though they can theoretically break someone. It's worth it to prevent typos like
```ts
el.autocapitalize: 'character';
```

**Comment by michaelwarren1106** [comment]
i wouldn’t mind making a PR if it’s relatively simple to just change the type. Like @Peeja said i’m sure there are other props like this whose types are less “strict” than the browser. would those need to be updated as well or just stick with this one since it’s known?

**Comment by RyanCavanaugh** [comment]
`HTMLElement.autocapitalize` is still declared as `string` in both TypeScript 5.1.3 and the current native build, so an invalid keyword such as `"character"` is accepted.

```ts
declare const element: HTMLElement;

// Expected: invalid autocapitalize keywords are rejected.
// Actual: HTMLElement.autocapitalize is string, so this is accepted.
element.autocapitalize = "character";
```

The browser accepts arbitrary strings and applies the attribute's invalid-value default (`"sentences"` in the tested browser), but a keyword union would catch misspellings such as `"character"` before they reach the platform.
